File: 2020/julia/AOC/src/day11.jl
module Day11 
import AOC.Parsers

const offsets = [CartesianIndex(-1, 0), CartesianIndex(1, 0),
				 CartesianIndex(0, -1), CartesianIndex(0, 1),
				 CartesianIndex(1, 1), CartesianIndex(-1, 1),
				 CartesianIndex(1, -1), CartesianIndex(-1, -1)]

function step_sim!(newboard, counter, tolerance, oldboard)
	for idx in CartesianIndices(oldboard) 
		if oldboard[idx] == 'L' && counter(oldboard, idx) == 0
			 newboard[idx] = '#'
		elseif oldboard[idx] == '#' && counter(oldboard, idx) >= tolerance
			newboard[idx] = 'L'
		else 
			newboard[idx] = oldboard[idx]
		end
	end
end

function run_sim(counter, tolerance, board)
	starting = 1
	ending = count(x -> x == '#', board)
	mutboard = copy(board)
	while starting - ending != 0
		step_sim!(mutboard, counter, tolerance, board)
		board, mutboard = mutboard, board
		starting, ending = ending, count(x -> x == '#', board)
	end
	ending
end

function count_adjacent(board, idx)
	count(offsets) do off 
		newidx = idx + off
		if checkbounds(Bool, board, newidx)
			board[newidx] == '#'
		else 
			false 
		end
	end
end

part1(;data=Parsers.day11()) = run_sim(count_adjacent, 4, data)

function run_length(board, off, idx)
	distance = 1
	while checkbounds(Bool, board, off*distance + idx)
		if board[off*distance + idx] == 'L'
			return false 
		elseif board[off*distance + idx] == '#'
			return true 
		end
		distance += 1
	end
	false
end

count_visible(board, idx) = count(off -> run_length(board, off, idx), offsets)

part2(;data=Parsers.day11()) = run_sim(count_visible, 5, data)

end

File: 2020/fsharp/AOC/Day11.fs
module AOC.Day11 

let directions = [|(-1, 0); (1, 0); (0, 1); (0, -1);
                  (-1, -1); (-1, 1); (1, -1); (1, 1)|]

let stepSim counter tolerance board = 
    Array2D.mapi (fun row col ele -> 
                  if ele = 'L' && (counter board row col) = 0 then
                    '#'
                  elif ele = '#' && (counter board row col) >= tolerance then
                    'L'
                  else ele) board

let countHash board = 
    board |> Seq.cast<char> |> Seq.filter ((=) '#') |> Seq.length

let runSim counter tolerance board = 
    let rec inner starting ending prev =
        if (starting - ending) = 0 then ending 
        else 
            let next = stepSim counter tolerance prev 
            inner ending (countHash next) next
    inner 1 (countHash board) board

let checkBounds row col board =
    ((row < Array2D.length1 board) && (row >= 0)
    && (col < Array2D.length2 board) && (col >= 0))

let part1 data = 
    let counter board row col =
        directions
        |> Seq.filter (fun (offrow, offcol) -> 
                    let (newrow, newcol) = (row + offrow, col + offcol)
                    if checkBounds newrow newcol board then board.[newrow, newcol] = '#'
                    else false)
        |> Seq.length
    runSim counter 4 data

let runLength (row, col) board (offrow, offcol) =
    let rec inner distance = 
        let newrow = offrow * distance + row
        let newcol = offcol * distance + col
        if not (checkBounds newrow newcol board) then false 
        elif board.[newrow, newcol] = 'L' then false 
        elif board.[newrow, newcol] = '#'  then true
        else inner (distance + 1)
    inner 1

let part2 data = 
    let counter board row col = 
        directions 
        |> Seq.filter (runLength (row, col) board)
        |> Seq.length
    runSim counter 5 data
